lib/chef/resource.rb

```diff
@@ -1274,31 +1274,48 @@ def self.resource_matching_short_name(short_name)
       end
     end
 
-    # Implement deprecated LWRP class
-    module DeprecatedLWRPClass
-      # @api private
-      def register_deprecated_lwrp_class(resource_class, class_name)
-        if Chef::Resource.const_defined?(class_name, false)
-          Chef::Log.warn "#{class_name} already exists!  Cannot create deprecation class for #{resource_class}"
-        else
-          deprecated_constants[class_name.to_sym] = resource_class
+    # @api private
+    def self.register_deprecated_lwrp_class(resource_class, class_name)
+      if Chef::Resource.const_defined?(class_name, false)
+        Chef::Log.warn "#{class_name} already exists!  Deprecation class overwrites #{resource_class}"
+        Chef::Resource.send(:remove_const, class_name)
+      end
+      resource_subclass = class_eval <<-EOM, __FILE__, __LINE__+1
```

Comment:
I also think it would be helpful to have some comments above this explaining why we want to subclass the anonymous `resource_class`.  IE, this is all necessary so that `Chef::Resource::LwrpClass` still works (until Chef 13)